core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -472,111 +369,77 @@ public BasicInvocationStage visitGetAllCommand(InvocationContext ctx, GetAllComm
-                  requestedKeys.add(key);
-               } else {
-                  if (trace) {
-                     log.tracef("Not doing a remote get for missing key %s since entry is "
-                                 + "mapped to current node (%s). Owners are %s",
-                           toStr(key), rpcManager.getAddress(), ch.locateOwners(key));
-                  }
-                  // Force a map entry to be created, because we know this entry is local
-                  entryFactory.wrapExternalEntry(ctx, key, null, EntryFactory.Wrap.WRAP_ALL, false);
-               }
+               Address primaryOwner = ch.locatePrimaryOwner(key);
+               requestedKeys.computeIfAbsent(primaryOwner, po -> new ArrayList<>()).add(key);
             }
          }
 
-         boolean missingRemoteValues = false;
-         if (!requestedKeys.isEmpty()) {
-            if (trace) {
-               log.tracef("Fetching entries for keys %s from remote nodes", requestedKeys);
-            }
+         if (requestedKeys.isEmpty()) {
+            return invokeNext(ctx, command);
+         }
 
-            Map<Object, InternalCacheEntry> justRetrieved = retrieveFromRemoteSources(
-                  requestedKeys, ctx, command.getFlagsBitSet());
-            Map<Object, InternalCacheEntry> previouslyFetched = command.getRemotelyFetched();
-            if (previouslyFetched != null) {
-               previouslyFetched.putAll(justRetrieved);
-            } else {
-               command.setRemotelyFetched(justRetrieved);
-            }
-            for (Object key : requestedKeys) {
-               if (!justRetrieved.containsKey(key)) {
-                  missingRemoteValues = true;
+         GlobalTransaction gtx = ctx.isInTxScope() ? ((TxInvocationContext) ctx).getGlobalTransaction() : null;
+         CompletableFutureWithCounter allFuture = new CompletableFutureWithCounter(requestedKeys.size());
+
+         for (Map.Entry<Address, List<Object>> pair : requestedKeys.entrySet()) {
+            List<Object> keys = pair.getValue();
+            ClusteredGetAllCommand clusteredGetAllCommand = cf.buildClusteredGetAllCommand(keys, command.getFlagsBitSet(), gtx);
+            rpcManager.invokeRemotelyAsync(Collections.singleton(pair.getKey()), clusteredGetAllCommand, defaultSyncOptions).whenComplete((responseMap, throwable) -> {
+               if (throwable != null) {
+                  allFuture.completeExceptionally(throwable);
+               }
+               Response response = getSingleSuccessfulResponseOrFail(responseMap, allFuture);
+               if (response == null) return;
+               Object responseValue = ((SuccessfulResponse) response).getResponseValue();
+               if (responseValue instanceof InternalCacheValue[]) {
+                  InternalCacheValue[] values = (InternalCacheValue[]) responseValue;
+                  int counterValue;
+                  synchronized (allFuture) {
```

Comment:
There are several remote invocations (this `invokeRemotelyAsync` is called in the for cycle), and this return handler can be invoked concurrently by many of them. As it modifies not only `allFuture.counter` (you could use AtomicInteger there) but also the context by wrapping the entries, we have to synchronize.